### ISSUE TITLE
在多卡上训练混淆矩阵没有合并

### ISSUE THREAD

**Issue by JingqingXu** [ultralytics/ultralytics#24796]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

gather_stats(self) 这个函数只在终端输出验证集数量时合并了不同卡的验证集，但在画混淆矩阵时没有将不同卡上的数据合并，只输出0卡上的数据，问题代码ultralytics/models/yolo/detect/val.py

### Environment

[ultralytics-disk-cache-hook] enabled, ultralytics=8.4.0, cache root: /tmp/ultralytics-disk-cache, image_disk_cache=True, dataset_meta_cache=True
New https://pypi.org/project/ultralytics/8.4.63 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.4.0 🚀 Python-3.10.19 torch-2.10.0.dev20251208+cu128 CUDA:11 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:10 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:9 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:8 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:7 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:6 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:5 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:4 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:3 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:2 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:1 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:0 (NVIDIA GeForce RTX 5090, 32109MiB)

### Minimal Reproducible Example

ultralytics/models/yolo/detect/val.py:222

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JingqingXu, thank you for reporting this issue with Ultralytics 🚀! This is an automated response to help gather the key details quickly, and an Ultralytics engineer will also assist you soon 😊

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 Bug Report, please provide a complete [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it. The file path and line reference are helpful, but we also need a small runnable reproduction showing the multi-GPU training/validation setup, exact command used, expected behavior, and actual confusion matrix output behavior across ranks.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please also upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify the issue is not already resolved in a newer version, since your report shows `8.4.0` and a newer release is available:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

Can you try the latest version of Ultralytics?

**Comment by JingqingXu** [comment]
> 你好。
> 
> 您能试用一下最新版本的 Ultralytics 吗？

我下载了ultralytics-8.4.65然后在/home/user/Minanconda//envs/yolo26/lib/python3.10/site-packages/ultralytics/models/yolo/detect/val.py中发现以下代码，这里明显没有合并不同卡上的混淆矩阵
def Gather_stats(self) -> None:
"""Gather stats from all GPU。"""
if RANK == 0:
Gather_stats = [None] * dist.get_world_size()
dist.gather_object(self.metrics.stats, Gather_stats, dst=0)
merged_stats = {key: [] for key in self.metrics.stats.keys()}
for stats_dict in Gathering_stats:
for key in merged_stats:
merged_stats[key].extend(stats_dict[key])
Gathered_jdict = [None] * dist.get_world_size()
dist.gather_object(self.jdict, gather_jdict, dst=0)
self.jdict = []
for jdict in gather_jdict:
self.jdict.extend(jdict)
self.metrics.stats = merged_stats
self._gather_image_metrics(self.metrics.box)
self.seen = len(self.dataloader.dataset) # 数据集中的图像总数
elif RANK > 0:
dist.gather_object(self.metrics.stats, None, dst=0)
dist.gather_object(self.jdict, None, dst=0)
self._gather_image_metrics(self.metrics.box)
self.jdict = []
self.metrics.clear_stats()

**Comment by Y-T-G** [comment]
Thanks for debugging. 

Can you test with:

```
pip install git+https://github.com/ultralytics/ultralytics@conf-mat-gather
```

**Comment by onuralpszr** [comment]
@JingqingXu hello could you able to test it ?

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24803: Gather confusion matrix across GPUs in DDP validation](https://github.com/ultralytics/ultralytics/pull/24803).

This update may resolve the issue by improving how the confusion matrix is handled during distributed validation:

- adds synchronization logic in `ultralytics/models/yolo/detect/val.py`
- when `plots=True` in DDP or multi-GPU validation, the confusion matrix is gathered across all processes
- the matrix is converted to a PyTorch tensor on the active device, reduced across workers, and restored on the main process
- this affects confusion matrix aggregation and plotting only, without changing the core metric computation flow

The result is that confusion matrix plots should now reflect results from all distributed workers instead of only a single process, which should make multi-GPU validation outputs more accurate.

Credit to @Y-T-G for the PR, with contributions from @onuralpszr and @glenn-jocher.

If you'd like to test the fix, you can install the latest `main` branch with:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Or you can wait for the next official PyPI release if preferred.

If possible, please check whether these changes resolve the issue in your setup and share the results. Thanks 🙏 for reporting this, and any further feedback is welcome if the issue still persists.
